### PR TITLE
[Hello] Add optional ROS driver, hook up Hector SLAM for base odometry

### PR DESCRIPTION
**Pull request by soumith**
## Effects: 

This PR brings in
-  significantly improved base odometry powered by a global LIDAR scan matcher, combined with a global optimizer that accounts for loop closures
- fixes some bugs in the planner that were making it stuck in-place in rare situations, because the grid map resolution, combined with navigation thresholds made it loop indefinitely

## Infrastructure

This PR adds:

- A ROS driver to
  - the base and joint {movement and statuses}
  - the lidar scans
  - [Hector SLAM](http://wiki.ros.org/hector_slam), a lidar-based SLAM with loop closure

The ROS driver was added for the base/joint/lidar because ROS generally wants to take full control of sensors -- and Hector SLAM is only available via ROS. The base/joint/lidar APIs dont really allow simultaneous usage among two processes.

The ROS driver is fully consistent with the non-ROS API. Userland should notice no difference.

### First commit: ROS drivers

You would want to review the first commit first, it's a large meaty commit.

- `launch.sh` adds in launching of ROS or non-ROS codepath via the `./launch.sh --ros` command-line argument
  - If running the ROS codepath, the user is expected to run `./launch_ros.sh` before-hand in a different terminal (as documented in the README). This launches some of the [stretch_core](https://github.com/hello-robot/stretch_ros/tree/master/stretch_core) ROS Services, and then launches hector_slam ROS Services.
- The ROS driver changes are isolated into four files (apart from `.launch` files):
  - `remote_hello_robot_ros.py` - this gives an API-compatible services as `remote_hello_robot.py`, but in the backend uses the ROS driver
  - `lidar_ros_driver.py` - this is the ROS driver that exposes lidar scans in the exact same format as original driver `lidar.py`
  - `stretch_ros_move_api.py` - this is the ROS driver that exposes move, base odometry, joint states, slam odometry
- `remote_hello_realsense.py` now has an optional `--ros` argument, which enables the ROS codepath for pulling LIDAR scans (and hence imports and uses `lidar_ros_driver.py` if this option is enabled)
- `goto` in `utils.py` which executes base movement for straightline trajectories has been generalized to work for both ROS and non-ROS backends

Other notes for this commit:
- We use the defaults for Hector SLAM (stretch_hector*.launch files), not trying to tune things yet. The defaults work great.

### Second commit:

- A subtle type-promotion bug identified during operation. If you give the XYT to `transform_global_to_base` as `[0, 0, 0]` then it treats it as a numpy int array, and the output transform ends up being wrong. The fix is simple, always cast it to float32 on entry

### Third commit:

The planner has lower resolution than the real world. By default, it has a planning grid with 5cm x 5cm tiles.
So, When the planner gets a given goal in world co-ordinates, it pulls a short-term goal, and it checks if the goal is within planner's resolution bounds.
However, this bounds-checking was incorrect because it used L2 Distance over the side of the grid, didn't account for diagonal / hypotenuse cases. I just simplified it by checking individual dimensions separately.

### CODE REVIEW

File: droidlet/lowlevel/hello_robot/remote/README.md [hunk 1]
```diff
@@ -20,6 +20,29 @@ popd
...
+# export LOCOBOT_IP=$(tailscale ip --4)
+
 ./launch.sh
+```
+
+### With ROS (i.e. better odometry using hector_slam, etc.)
+
+```bash
+# in fresh terminal
+deactivate
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/README.md:37]: Why do we need this?

File: droidlet/lowlevel/hello_robot/remote/README.md [hunk 2]
```diff
@@ -20,6 +20,29 @@ popd
...
+./launch_ros.sh
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/README.md:38]: extra steps for launching ROS - should we discuss? I think if its only one ROS launch file maybe it could go into submodule or MRP 

doesnt need to be a change for this PR though

File: droidlet/lowlevel/hello_robot/remote/launch_ros.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/env bash
+set -e
+
+if $(which python | grep fairo >/dev/null); then
+    echo "droidlet virtualenv needs to be deactivated before running this script. Run 'deactivate' in the terminal before running this"
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/launch_ros.sh:5]: will this cause issues?

File: droidlet/lowlevel/hello_robot/remote/lidar_ros_driver.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import math
+import sys
+import threading
+import time
+import traceback
+import rospy
+from sensor_msgs.msg import LaserScan
+
+
+class Lidar:
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/lidar_ros_driver.py:10]: Should this share base class hierarchy with non-ROS version?

Comment (cpaxton) [reply droidlet/lowlevel/hello_robot/remote/lidar_ros_driver.py:10]: It would be nice to provide a commo interface

File: droidlet/lowlevel/hello_robot/remote/remote_hello_realsense.py [hunk 5]
```diff
@@ -40,8 +39,14 @@
 class RemoteHelloRealsense(object):
     """Hello Robot interface"""
 
-    def __init__(self, bot):
+    def __init__(self, bot, use_ros=False):
         self.bot = bot
+        if use_ros:
+            self.use_ros = True
+            from droidlet.lowlevel.hello_robot.remote.lidar_ros_driver import Lidar
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_realsense.py:46]: So they are the same - might be nice to have Lidar ABC?

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot.py [hunk 6]
```diff
@@ -75,6 +75,9 @@ def _check_battery(self):
         val_in_range("CPU Temp", p.status["cpu_temp"], vmin=15, vmax=80)
         print(Style.RESET_ALL)
 
+    def pull_status(self):
+        self._robot.pull_status()
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_robot.py:79]: short comment to document function? not clear what this does to me

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot.py [hunk 7]
```diff
@@ -233,10 +236,23 @@ def go_to_relative(self, xyt_position):
...
+            or robot.base.left_wheel.status["is_moving"]
+        )
+        right_wheel_moving = (
+            robot.base.right_wheel.status["is_moving_filtered"]
+            or robot.base.right_wheel.status["is_moving"]
+        )
+        is_moving = left_wheel_moving or right_wheel_moving
+        return is_moving
+
     def is_busy(self):
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_robot.py:256]: Should this include arm motion?

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+
+Pyro4.config.SERIALIZER = "pickle"
+Pyro4.config.SERIALIZERS_ACCEPTED.add("pickle")
+Pyro4.config.ITER_STREAMING = True
+
+
+# #####################################################
+@Pyro4.expose
+class RemoteHelloRobot(object):
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:28]: Seems like a good amount of duplicate code here

Comment (cpaxton) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:28]: maybe we should just create a base class for RemoteHelloRobot?

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        return self._robot.get_slam_pose()
+        # second Best (from lidar scanning)
+        # return self._robot.get_scan_matched_pose()
+        # Worst, from wheel encoder only
+        # return self._robot.get_odom()
+
+    def get_slam_pose(self):
+        return self._robot.get_slam_pose()
+
+    def pull_status(self):
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:105]: not implemented - raise notimplementederror?

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+    def get_tilt(self):
+        return self._robot.get_joint_state("head_tilt")
+
+    def set_pan(self, pan):
+        self._robot.send_command("joint_head_pan", pan)
+
+    def set_tilt(self, tilt):
+        self._robot.send_command("joint_head_tilt", tilt)
+
+    def reset_camera(self):
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:125]: description - resets camera to home position?

File: droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+from control_msgs.msg import FollowJointTrajectoryGoal
+from trajectory_msgs.msg import JointTrajectoryPoint
+from geometry_msgs.msg import PoseStamped, Pose2D, PoseWithCovarianceStamped
+from nav_msgs.msg import Odometry
+import hello_helpers.hello_misc as hm
+from tf.transformations import euler_from_quaternion
+import collections
+import numpy as np
+
+class MoveNode(hm.HelloNode):
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py]: I don't fully understand why we need this class - maybe it would make sense to add some comments?

File: droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+                self._command = None                
+            if command is not None:
+                joint_name, increment = command
+                self._send_command(joint_name, increment)
+            rate.sleep()
+
+    def start(self):
+        hm.HelloNode.main(self, 'fairo_hello_proxy', 'fairo_hello_proxy', wait_for_first_pointcloud=False)
+        self._thread = threading.Thread(target=self.background_loop, daemon=True)
+        self._thread.start()
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py:165]: Remove dead code?

File: droidlet/lowlevel/hello_robot/remote/utils.py [hunk 13]
```diff
@@ -238,20 +223,12 @@ def goto(
...
-            )
-            right_wheel_moving = (
-                robot.base.right_wheel.status["is_moving_filtered"]
-                or robot.base.right_wheel.status["is_moving"]
-            )
-            is_moving = left_wheel_moving or right_wheel_moving
+            is_moving = robot.is_base_moving()
 
     # second rotate by theta2
     theta_2 = np.sign(theta_2) * (abs(theta_2) % radians(360))
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/utils.py:234]: what are theta_1 vs theta_2?

File: droidlet/lowlevel/locobot/remote/slam_service.py [hunk 14]
```diff
@@ -91,7 +92,7 @@ def add_obstacle(self, location, in_map=False):
         self.map_builder.add_obstacle(location)
 
     def update_map(self):
-        pcd = self.robot.get_current_pcd()[0]
+        pcd = safe_call(self.robot.get_current_pcd)[0]
```
Comment (cpaxton) [inline droidlet/lowlevel/locobot/remote/slam_service.py:95]: What is safe_cal for? why is this necessary?

File: droidlet/lowlevel/hello_robot/remote/utils.py [hunk 15]
```diff
@@ -196,7 +196,7 @@ def goto(
         robot.rotate_by(theta_1)
         robot.push_command()
         robot.pull_status()
-        time.sleep(0.1)
+        time.sleep(1)
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/remote/utils.py:199]: any reason for these sleeps? should there be a generic "wait" for a status/command instead?

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py [hunk 16]
```diff
@@ -229,11 +229,11 @@ def move_relative(self, xyt_positions, blocking=True):
             xyt_positions = [xyt_positions]
         for xyt in xyt_positions:
             self.nav_result.wait()
-            self.nav_result = safe_call(self.nav.go_to_relative, xyt)
+            self.nav_result = safe_call(self.nav.go_to_relative, xyt, 10000000000, distance_threshold, angle_threshold)
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/hello_robot_mover.py]: strange hard coded value here

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py [hunk 17]
```diff
@@ -229,11 +229,11 @@ def move_relative(self, xyt_positions, blocking=True):
...
             if blocking:
                 self.nav_result.wait()
 
-    def move_absolute(self, xzt_positions, blocking=True):
+    def move_absolute(self, xzt_positions, blocking=True, distance_threshold=None, angle_threshold=None):
         """Command to execute a move to an absolute position.
 
         It receives positions in canonical world coordinates and converts them to pyrobot's coordinates
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/hello_robot_mover.py:245]: assuming None means that these are not used by default - but in that case how does it ever stop? does this have a timeout?

Comment (cpaxton) [reply droidlet/lowlevel/hello_robot/hello_robot_mover.py:245]: None might not be a good default if so

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py [hunk 18]
```diff
@@ -250,7 +250,7 @@ def move_absolute(self, xzt_positions, blocking=True):
             logging.info("Move absolute in canonical coordinates {}".format(xzt))
             self.nav_result.wait()
             robot_coords = base_canonical_coords_to_pyrobot_coords(xzt)
-            self.nav_result = self.nav.go_to_absolute(robot_coords)
+            self.nav_result = self.nav.go_to_absolute(robot_coords, 10000000000, distance_threshold, angle_threshold)
```
Comment (cpaxton) [inline droidlet/lowlevel/hello_robot/hello_robot_mover.py]: is 10000000 the timeout?

File: droidlet/lowlevel/locobot/remote/planning_service.py [hunk 19]
```diff
@@ -76,27 +78,31 @@ def get_short_term_goal(self, robot_location, goal, step_size=25):
             target_goal = (stg_real[0], stg_real[1], rotation_angle)
         return target_goal
 
-    def goal_within_threshold(self, robot_location, goal):
-        distance = np.linalg.norm(np.array(robot_location[:2]) - np.array(goal[:2]))
-        # in metres. map_resolution is the resolution of the SLAM's 2D map, so the planner can't
-        # plan anything lower than this
-        threshold = (float(self.map_resolution) - 1e-10) / 100.0
+    def goal_within_threshold(self, robot_location, goal, threshold=None, angle_threshold=None):
```
Comment (cpaxton) [inline droidlet/lowlevel/locobot/remote/planning_service.py:82]: description would be good - although this one seems pretty self explanatory

Comment (cpaxton) [reply droidlet/lowlevel/locobot/remote/planning_service.py:82]: goal and location are the same type?

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+        # at 90, the correction is +0.11526719 + -0.091491526943
+        # linear interpolate the correction of 0.023775
+        interp_correction = 0.11526719 * abs(head_tilt) / radians(90)
+        # print('interp_correction', interp_correction)
+
+        camera_transform[2, 3] += -0.091491526943 + interp_correction
+
+        return camera_transform
+
+    def get_base_state(self):
```
Comment (theophilee) [inline droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:95]: The `yaw` is in `[-3.14, 3.14]` when using Hector SLAM and in `[0, 6.28]` when not, maybe we could make it consistent?

File: droidlet/lowlevel/hello_robot/remote/README.md (hunk 1 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/README.md:39]: this is to make sure we aren't in any kind of virtualenv / pipenv. The ROS launch commands need to use system python as-is.

File: droidlet/lowlevel/hello_robot/remote/README.md (hunk 2 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/README.md:40]: Will move everything to MRP in an upcoming PR

File: droidlet/lowlevel/hello_robot/remote/launch_ros.sh (hunk 3 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/launch_ros.sh:5]: not really. it's a check to make sure issues aren't caused by accidentally forgetting to do `deactivate` or having a virtualenv active

File: droidlet/lowlevel/hello_robot/remote/remote_hello_realsense.py (hunk 5 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_realsense.py:46]: done

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py (hunk 9 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:106]: This is intentionally intended to be empty, because we want to match the API between this and the non-ROS class

Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:106]: we dont need to do a pull_status with the ROS driver

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py (hunk 10 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:126]: done

File: droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py (hunk 11 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py]: done. yes it's the Move API but via ROS.
Once the robot is driven via the ROS stuff (i.e. to use hector_slam), then we cant use the move API directly anymore and are forced to control the robot via ROS as well

File: droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py (hunk 12 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/stretch_ros_move_api.py:165]: done

File: droidlet/lowlevel/hello_robot/remote/utils.py (hunk 13 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/utils.py:234]: they're described in the comments in the file.

File: droidlet/lowlevel/locobot/remote/slam_service.py (hunk 14 above)
Comment (soumith) [reply droidlet/lowlevel/locobot/remote/slam_service.py:95]: needed for making pyro4 fail-fast in case of a downstream service error (instead of waiting to timeout).
All the safe_call business will go away once we move to zerorpc.

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py (hunk 16 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/hello_robot_mover.py]: removed

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py [hunk 21]
```diff
@@ -217,7 +217,7 @@ def reset_camera(self):
         """reset the camera to 0 pan and tilt."""
         return self.bot.reset()
 
-    def move_relative(self, xyt_positions, blocking=True):
+    def move_relative(self, xyt_positions, blocking=True, distance_threshold=None, angle_threshold=None):
         """Command to execute a relative move.
 
         Args:
```
[the remark this answers is not in the document]
Comment (soumith) [reply droidlet/lowlevel/hello_robot/hello_robot_mover.py:225]: done

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py (hunk 17 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/hello_robot_mover.py:245]: None will make it default to the planner's default values, which are dynamic based on the map resolution.
For example, the default planner's distance threshold is:
```
threshold = (float(self.map_resolution) - 1e-10) / 100.0
```

Comment (soumith) [reply droidlet/lowlevel/hello_robot/hello_robot_mover.py:245]: this makes sense, because the planner cant plan a path with a resolution lower than it's map resolution anyways

File: droidlet/lowlevel/hello_robot/hello_robot_mover.py (hunk 18 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/hello_robot_mover.py]: removed it

File: droidlet/lowlevel/locobot/remote/planning_service.py (hunk 19 above)
Comment (soumith) [reply droidlet/lowlevel/locobot/remote/planning_service.py:82]: goal and location are of same type, indeed.

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py (hunk 20 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:95]: done. added a `math.pi` to the yaw in this case, so that it's consistent to `[0, 6.28]`.

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py [hunk 22]
```diff
@@ -197,13 +205,22 @@ def go_to_relative(self, xyt_position):
...
+                print(e)
+                print(traceback.format_exc())
+                self._done = True
+                raise e
         return status
 
     def is_moving(self):
         return not self._done
 
+    def is_busy(self):
```
[the remark this answers is not in the document]
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:251]: yes, when we add in the API to move the Arm

File: droidlet/lowlevel/hello_robot/remote/utils.py (hunk 15 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/utils.py:199]: there isn't a generic `wait` available in the base API for stretch.
But I spoke to the Hello folks a few days back and there's a better way to handle this than the sleep.
Will implement in a follow-up PR

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py (hunk 8 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot_ros.py:29]: will open an issue for it.

File: droidlet/lowlevel/hello_robot/remote/remote_hello_robot.py (hunk 7 above)
Comment (soumith) [reply droidlet/lowlevel/hello_robot/remote/remote_hello_robot.py:265]: yes, but after we implement the arm capabilities

**Review by cpaxton — APPROVED** [review]
